packages/databricks-vscode/src/vscode-objs/StateStorage.ts

```diff
@@ -1,73 +1,100 @@
 import {randomUUID} from "crypto";
 import {ExtensionContext} from "vscode";
 
-export class StateStorage {
-    constructor(private context: ExtensionContext) {}
+/* eslint-disable @typescript-eslint/naming-convention */
+type KeyInfo<V> = {
+    location: "global" | "workspace";
+    defaultValue?: V;
+    getter?: (storage: StateStorage, value: V | undefined) => V | undefined;
+    setter?: (storage: StateStorage, value: V | undefined) => V | undefined;
+};
 
-    get fixedRandom() {
-        let randomNum = this.context.globalState.get<number>(
-            "databricks.fixedRandom"
-        );
-        if (!randomNum) {
-            randomNum = Math.random();
-            this.context.globalState.update(
-                "databricks.fixedRandom",
-                randomNum
-            );
-        }
-        return randomNum;
-    }
+function withType<V>() {
+    return function <D extends KeyInfo<V>>(data: D) {
```

Comment:
Why nested functions? Can outer function already return data as `KeyInfoWithType<V>`? Later in the `storage.get` you won't need to use that `as` anymore.

Reply:
This is related to how I wanted getters to work. Replied [here](https://github.com/databricks/databricks-vscode/pull/913#discussion_r1368593091)